vault/ui_custom_messages/entry.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package uicustommessages
+
+import (
+	"errors"
+
+	"github.com/hashicorp/go-uuid"
+)
+
+// Entry is a struct that contains a map of custom message ID to Message struct
+// that can be marshalled and unmarshalled using the JSON encoding to/from a
+// slice of byte to translate to a logical.StorageEntry struct. An Entry
+// consists of data for a particular namespace.
+type Entry struct {
+	// Messages is the map of custom message ID to Message struct.
+	Messages map[string]Message `json:"messages"`
+}
+
+// FindMessages searches through all of the custom messages in the receiver
+// Entry struct and only returns those that match the criteria set in the
+// provided FindFilter struct.
+func (e *Entry) FindMessages(filter FindFilter) []Message {
+	result := make([]Message, 0)
+
+	for _, message := range e.Messages {
+		if message.Matches(filter) {
+			result = append(result, message)
+		}
+	}
+
+	return result
+}
+
+// CreateMessage adds a custom message in the receiver Entry struct using the
+// provided Message struct to populate its properties. If the either the
+// start/end times are invalid or the maximum number of messages already exists,
+// then the message is not added.
+func (e *Entry) CreateMessage(message *Message) error {
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return err
+	}
+
+	if !message.ValidateStartAndEndTimes() {
+		return errors.New("message start time must occur before end time")
+	}
+
+	if !message.ValidateMessageType() {
+		return errors.New("unrecognized message type")
+	}
+
+	// This condition should be evaluated last, because if anything else was to
+	// prevent the creation of the message, there's no use bringing up the
+	// limit.
+	if len(e.Messages) >= MaximumMessageCountPerNamespace {
+		return errors.New("maximum number of messages already exist")
+	}
+
+	message.ID = uuid
+
+	if e.Messages == nil {
+		e.Messages = make(map[string]Message)
+	}
+	e.Messages[uuid] = *message
+
+	return nil
+}
+
+// UpdateMessage updates the Message struct stored in the receiver's Messages
+// map with the provided Message struct. If the start/end times are invalid, the
+// message is not updated. The Messages map is not changed if it does not
+// contain the key message.ID.
+func (e *Entry) UpdateMessage(message *Message) (bool, error) {
+	if e.Messages == nil {
+		e.Messages = make(map[string]Message)
+	}
+
+	if _, ok := e.Messages[message.ID]; !ok {
+		// Special case: returning nil, nil indicates that the message with the
```

Comment:
Typo: `false, nil` instead of `nil, nil`

However, I wonder if we should return a custom message to indicate not found, like `ErrMessageNotFound`? We do a similar pattern in the cache of Agent/Proxy with `ErrCacheItemNotFound`.

Reply:
Good catch on the stale comment. I think what you are proposing (wrt the error) is already being done in the `(*SystemBackend).handleUpdateCustomMessage` method. It makes the determination based on the values returned by `(*uicustommessages.Manager).UpdateMessage` that if the message wasn't found, it injects `logical.ErrCustomMessageNotFound` as an error along with the nil `logical.Response`.

Reply:
It's returning that as an API response, yeah, but I think that it makes sense to me to do so for internal functions, too. It would avoid needing comments like this, and it avoids patterns like returning `nil, nil` which can be confusing

Reply:
I took a closer look and there's nothing stopping emitting the `logical.ErrCustomMessageNotFound` error from this function and it reduces some extra evaluations to detect the condition higher up the stack.

Reply:
Hmm, I guess that makes sense! To me, it does make sense for this error to live in the ui messages packages, but I guess at the API layer that doesn't stop us emitting it.

Reply:
Nick jumped on this too and reiterate his distain of the `return nil, nil` to signify Not Found. So, I've replaced the error with a more generic one: `logical.ErrNotFound` and I wrap that error using `fmt.Errorf()`. This allows using the `errors.Is` function further up the stack to detect if an error returned wraps `logical.ErrNotFound`.

Reply:
This is a little personal preference but to me it feels nicer to have a custom error for this kind of thing, as opposed to a more generic one. That's the pattern we've used elsewhere and it lets us be more specific that we're specifically looking for a 'message not found' kind of error